File: Projects/codechef/DDIMMST.java
package codechef;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.Comparator;
import java.util.PriorityQueue;
import java.util.Scanner;

class DDIMMST {
    private static int distance(Point p1, Point p2) {
        final int D = p1.x.length;
        int w = 0;
        for (int i = 0; i < D; i++) {
            w += Math.abs(p1.x[i] - p2.x[i]);
        }
        return w;
    }

    public static void main(String[] args) throws Exception {
        final Field queue = PriorityQueue.class.getDeclaredField("queue");
        queue.setAccessible(true);
        final Method siftUp = PriorityQueue.class.getDeclaredMethod("siftUp", int.class, Object.class);
        siftUp.setAccessible(true);
        final Scanner stdin = new Scanner(new BufferedReader(new InputStreamReader(System.in)));
        final int N = stdin.nextInt();
        final int D = stdin.nextInt();
        final Point[] p = new Point[N];
        for (int n = 0; n < N; n++) {
            final int[] x = new int[D];
            for (int d = 0; d < D; d++) {
                x[d] = stdin.nextInt();
            }
            p[n] = new Point(x);
        }
        int weight = 0;
        final PriorityQueue<Point> pool = new PriorityQueue<>(N, Comparator.comparingInt(Point::w).reversed());
        pool.addAll(Arrays.asList(p));
        final Object[] array = (Object[]) queue.get(pool);
        while (!pool.isEmpty()) {
            final Point s = pool.remove();
            Point t = null;
            for (int n = 0; n < N && (t = (Point) array[n]) != null; n++) {
                final int w = distance(s, t);
                if (w > t.w) {
                    t.set(w);
                    siftUp.invoke(pool, n, t);
                }
            }
            weight += s.w;
        }
        System.out.println(weight);
        stdin.close();
    }

    private static class Point {
        private final int[] x;
        private int w = 0;

        public Point(int[] x) {
            this.x = x;
        }

        public int w() {
            return this.w;
        }

        public void set(int w) {
            this.w = w;
        }
    }
}

File: Projects/codechef/DDIMMST.kt
package codechef

import java.io.*

internal object DDIMMST {
    private fun distance(p1: Point, p2: Point?): Int {
        val D = p1.x.size
        var w = 0
        for (i in 0 until D) {
            w += Math.abs(p1.x[i] - p2!!.x[i])
        }
        return w
    }

    @Throws(Exception::class)
    fun main(args: Array<String?>?) {
        val queue: Field = PriorityQueue::class.java.getDeclaredField("queue")
        queue.setAccessible(true)
        val siftUp: Method = PriorityQueue::class.java.getDeclaredMethod("siftUp", Int::class.javaPrimitiveType, Object::class.java)
        siftUp.setAccessible(true)
        val stdin = Scanner(BufferedReader(InputStreamReader(System.`in`)))
        val N: Int = stdin.nextInt()
        val D: Int = stdin.nextInt()
        val p = arrayOfNulls<Point>(N)
        for (n in 0 until N) {
            val x = IntArray(D)
            for (d in 0 until D) {
                x[d] = stdin.nextInt()
            }
            p[n] = Point(x)
        }
        var weight = 0
        val pool: PriorityQueue<Point> = PriorityQueue(N, Comparator.comparingInt { obj: Point -> obj.w() }.reversed())
        pool.addAll(Arrays.asList(p))
        while (!pool.isEmpty()) {
            val s: Point = pool.remove()
            var t: Point? = null
            var n = 0
            while (n < N && queue.get(pool)[n] as Point?. also { t = it } != null) {
                val w = distance(s, t)
                if (w > t!!.w) {
                    t!!.set(w)
                    siftUp.invoke(pool, n, t)
                }
                n++
            }
            weight += s.w
        }
        System.out.println(weight)
        stdin.close()
    }

    private class Point(val x: IntArray) {
        var w = 0
        fun w(): Int {
            return w
        }

        fun set(w: Int) {
            this.w = w
        }
    }
}